# 02_tweet_analysis
For the RootCause/Corridor Labs collab, we want to do a range of analyses on the tweets we have collected.  
This notebook contains said analyses, ordered in chapters. The chapters are as follows: 
1. Qatar-related tweets: How has (simple) **sentiment** evolved over time and what can we learn from it?
2. **Tweets in general. Production of a descriptive overview of what the tweets we have collected contain, sorted by popularity/frequency.**
3. MERS tweets. Can we find any interesting patterns in world-cup-tangential tweets mentioning 'MERS'/'camel flu'?
4. General investigations into misinformation etc. Can we find origins and evolutions about the Saudi X Ray image? 

NL, 22/12/22  
NL, 23/12/22 -- re-jigging so as to not overcount retweets for non-original tweets

## IMPORTS

In [36]:
import os
import json
import pandas as pd
import numpy as np
from tqdm import tqdm

import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px
import plotly.graph_objects as go
import plotly.io as pio
from wordcloud import WordCloud, STOPWORDS, ImageColorGenerator

## PATHS & CONSTANTS

In [57]:
DATES = ['2022_11_26', '2022_12_04', '2022_12_14', '2022_12_18']

TWEETS_PATH = '/home/nikloynes/projects/world_cup_misinfo_tracking/data/tweets_clean/'
ENTITIES_PATH = '/home/nikloynes/projects/world_cup_misinfo_tracking/data/tweets_meta/'
EXPORT_PATH = '/home/nikloynes/projects/world_cup_misinfo_tracking/data/exports/tweets/'

## INIT

In [3]:
pio.renderers.default='notebook'

In [34]:
pd.set_option('display.max_rows', 100)

## THE THING!

### TWEETS
Now, in chapter 2, we shift our attention to the general analysis of *all* the tweets we've collected. 

1. Who are the top users - per chunk and in total? We will find this out by iterating over tweets; retrieving user ids and adding/appending their likes/RTs/replies
2. What are the most frequent entities -- and what are some unusual, and perhaps interesting entities? For this we're leveraging Twitter's self-supplied `entities` info, which we've saved for each file already.
3. What were the top pieces of content (tweets, URLs, domains, videos, images)

### USERS

In [4]:
tweet_files = os.listdir(TWEETS_PATH)
tweet_files.sort()

In [5]:
tweet_files.remove('tweets_2022_12_18-16_00_01-old_broken.json')

In [18]:
all_users = {}

**Extracting core user-level metrics from our samples of tweets**

Note: we are only counting retweets if a user was the original author of a tweet (or quote tweet). We are not counting retweets for a pure retweet, as this would count all instances of a retweet of the original tweet - not those of the particular instance of a given users' retweet.  

In [19]:
tweets = []
for file in tweet_files:
    target_date = file[7:17]
    with open(TWEETS_PATH+file, 'r') as infile:
        for line in tqdm(infile):
            tmp = json.loads(line)
            # all users
            if tmp['author_id'] not in all_users.keys():
                all_users[tmp['author_id']] = {'tweet_count' : 1}
                # make sure we only count retweets if we're looking at a
                # first-author tweet.
                if not ('referenced_tweets' in tmp.keys() and 'retweeted' in [x['type'] for x in tmp['referenced_tweets']]): 
                    all_users[tmp['author_id']]['retweet_count'] = tmp['public_metrics']['retweet_count']
                else:
                    all_users[tmp['author_id']]['retweet_count'] = 0
                all_users[tmp['author_id']]['reply_count'] = tmp['public_metrics']['reply_count'] 
                all_users[tmp['author_id']]['like_count'] = tmp['public_metrics']['like_count']
                all_users[tmp['author_id']]['quote_count'] = tmp['public_metrics']['quote_count']
                all_users[tmp['author_id']][f'in_{target_date}'] = True

                # user-metrics (these are no generic, not specific to our samples)
                all_users[tmp['author_id']]['user_name'] = tmp['user']['username']
                all_users[tmp['author_id']]['follower_count'] = tmp['user']['public_metrics']['followers_count']
                all_users[tmp['author_id']]['following_count'] = tmp['user']['public_metrics']['following_count']

            else:
                all_users[tmp['author_id']]['tweet_count'] += 1
                if not ('referenced_tweets' in tmp.keys() and 'retweeted' in [x['type'] for x in tmp['referenced_tweets']]): 
                    all_users[tmp['author_id']]['retweet_count'] += tmp['public_metrics']['retweet_count']
                all_users[tmp['author_id']]['reply_count'] += tmp['public_metrics']['reply_count'] 
                all_users[tmp['author_id']]['like_count'] += tmp['public_metrics']['like_count']
                all_users[tmp['author_id']]['quote_count'] += tmp['public_metrics']['quote_count']
                all_users[tmp['author_id']][f'in_{target_date}'] = True

2264it [00:00, 61270.62it/s]
25719it [00:00, 68098.35it/s]
43888it [00:00, 68367.54it/s]
50354it [00:00, 70169.70it/s]
55748it [00:00, 69864.92it/s]
61778it [00:00, 68370.39it/s]
11487it [00:00, 66994.23it/s]
13194it [00:00, 64540.25it/s]
86704it [00:01, 65349.47it/s]
81108it [00:01, 65007.17it/s]
47279it [00:00, 66038.47it/s]
22075it [00:00, 65516.20it/s]
99529it [00:01, 58802.86it/s]
92643it [00:01, 58211.41it/s]
60033it [00:01, 59871.78it/s]
87739it [00:01, 62054.16it/s]
85849it [00:01, 61047.43it/s]
62355it [00:01, 60230.19it/s]
41891it [00:00, 61215.26it/s]
499it [00:00, 60287.99it/s]
31523it [00:00, 67250.85it/s]
1966it [00:00, 63807.24it/s]
15931it [00:00, 64131.39it/s]
120887it [00:01, 63563.36it/s]
159345it [00:02, 63495.73it/s]
160985it [00:02, 64316.83it/s]
185044it [00:02, 62333.73it/s]
156847it [00:02, 62694.98it/s]
44726it [00:00, 63541.99it/s]
267369it [00:04, 62014.93it/s]
527141it [00:08, 61069.86it/s]
442495it [00:07, 60557.07it/s]
470338it [00:07, 61591.04it/s]
99150

##

**Top-level counts**

In [20]:
print(f'N unique users in total: {len(all_users)}')

N unique users in total: 1852731


In [25]:
for date in DATES:
    print(f"N users for chunk {date}: {len([all_users[x][f'in_{date}'] for x in all_users.keys() if f'in_{date}' in all_users[x].keys()])}")

N users for chunk 2022_11_26: 307116
N users for chunk 2022_12_04: 279883
N users for chunk 2022_12_14: 517309
N users for chunk 2022_12_18: 1044254


In [26]:
all_users_df = pd.DataFrame(all_users).transpose().reset_index().rename(columns={'index' : 'user_id'})

In [28]:
# recode NaNs in `in_{date}` as False
for date in DATES:
    all_users_df[f'in_{date}'] = all_users_df[f'in_{date}'].fillna(False)

In [ ]:
all_users_df['interaction_count'] = all_users_df['like_count'] + all_users_df['quote_count'] + all_users_df['reply_count'] + all_users_df['retweet_count']

In [46]:
# re-order our columns
all_users_df = all_users_df[[
    'user_id',
    'user_name',
    'follower_count',
    'following_count',
    'tweet_count',
    'retweet_count',
    'reply_count',
    'like_count', 
    'quote_count',
    'interaction_count',
    'in_2022_11_26',
    'in_2022_12_04',
    'in_2022_12_14',
    'in_2022_12_18'
    ]]

In [47]:
prolific_users_df = all_users_df.loc[
    (all_users_df['in_2022_11_26']==True) &
    (all_users_df['in_2022_12_04']==True) &
    (all_users_df['in_2022_12_14']==True) &
    (all_users_df['in_2022_12_18']==True)      
]

In [48]:
np.mean(all_users_df['retweet_count'])

1.5220061627942751

In [49]:
np.mean(all_users_df['like_count'])

11.532634796956494

In [50]:
np.mean(all_users_df['reply_count'])

0.38381070970367526

In [51]:
np.mean(all_users_df['interaction_count'])

13.601833185713414

Some takeaways:
- We have a total of `1,852,731` unique users in our sample; out of a total of `3,844,895` tweets.
- This means that the mean number of tweets per user is slightly above `2` per user; the median is `1`.
- There are a total of `10,101` users which appear in all 4 samples.
- It is apparent that there is quite a difference in the number of tweet interactions based on how tweets were collected. Because of some issues with the server hosting this entire project, I had to collect 2 of the tweet chunk samples after the fact (1-2 days) using the *search* rather than the *streaming* method. In the search method; where tweets had a while to gather attention, our interaction numbers are much higher as compared to the streaming method, where we write a tweet to disk within seconds of it being published. If we're interested in getting complete, post-hoc interaction numbers for all tweets; we can do this via the *get_tweets* endpoint on the API (by supplying chunks of tweet ids which we've recorded). As far as I'm aware we aren't restricted by the 1-week time limit for this, but obviously the 2-million monthly limit applies, so it would have to happen via a different Twitter account's API credentials if it has to occur soon, or in January via Nik's Twitter.    
- As noted above, we're only looking at retweets of original and quote tweets; not pure retweets. The mean number of retweets per tweet is approx `1.5` -- but this is likely distorted given the issue outlined above. 
- Our mean number of likes per user is approx `11.5`. 
- Our mean number of replies per users is approx `0.4`.
- The mean number of any interaction per user is approx `13`.

#### User 'League Tables'

We will no look at some core metrics of interest, and display tables reflecting the **Top 100** users ranked based on that metric. I will leave any interpretation of this to the reader.

1. Any interaction

In [53]:
all_users_df.sort_values('interaction_count', ascending=False).reset_index()[:100]

index              user_id        user_name follower_count  \
0    106793            138372303     FIFAWorldCup       11173420   
1   1073074             44196397         elonmusk      122510051   
2    296818             18091004           ESPNFC        6958316   
3    329417            330262748   FabrizioRomano       12796954   
4   1055790             44409004          shakira       53305430   
5    257908           1561123663       brfootball        5967973   
6    356251  1138458175663988738          PopBase         467303   
7    354605            627673190  ChampionsLeague       45779558   
8   1472143            266135518              KSI        8560920   
9   1710867             14123683      andy_murray        3563717   
10  1316018   805532293951606785   MattWallace888         786600   
11   488317              2557521             espn       43696497   
12   306635             89452962           ESPNUK         655573   
13  1627288             57606154    OfficialBHAFC         712265   
14   645890           3324689194   TheEuropeanLad         468941   
15    54114            119593082        FOXSoccer        1436531   
16   238527             27489253     AVFCOfficial        2206526   
17   780824           4429003533         PopCrave        1202948   
18  1643404             43152482    AlejandroSanz       19800789   
19   167557  1303026765737218049        StokeyyG2         256585   
20   335777               890891   BleacherReport       13219827   
21  1559476            216299334      piersmorgan        8337074   
22   344665           4766665333      Culturecams          40529   
23   358568             26809005             goal        7149176   
24    65219   896550698543874049          UTDTrey         391678   
25   229445             69962890   2010MisterChip        3478499   
26   141738            435225922       sportbible        1835349   
27   148439           2333988375      lnstantFoot        1501308   
28    42996             39692424     AlertaNews24        1225927   
29  1662041             14130366     sundarpichai        5054964   
30  1004143            279581853     AchrafHakimi        1473974   
31  1387028           3437532637  myogiadityanath       22984910   
32   854770  1447233497731674117  AlbicelesteTalk         152174   
33    86532  1419145241018445825       _FaridKhan          38300   
34    84613            462312323  Football__Tweet         760537   
35   601606            533085085              M10       26308719   
36   277070            571964518    TrollFootball        2417482   
37   595911            427089628       lexfridman        2354327   
38  1006820           2192335506     TeamCRonaldo         287111   
39   481155             34613288          Arsenal       20756694   
40  1019814   973338313661861889  btschartsdailys         726344   
41  1801263            135421739        sachin_rt       38256550   
42   102523            385473825          Squawka        1219522   
43    34008           1580762162  chidajunsei3163         103445   
44   206942  1269288366954237952      mocnknights            679   
45   230144   953378142198161409          choquei        2515110   
46  1426033             29873662            MKBHD        6018521   
47   384471            343627165    premierleague       37699137   
48  1505537  1520612280643162112    DashDobrofsky         160402   
49    28487             15332636        talkSPORT        2078304   
50  1179701             26257166     SportsCenter       41400044   
51   176966             19583545              LFC       22907135   
52   165052            771606164        ActuFoot_        5361714   
53   325208           4488217998    InvictosSomos        1050462   
54   141168            265902729         BBCSport       10631968   
55   229951           2600210567              433        3288602   
56  1419836             92724677   virendersehwag       23152722   
57  1293148           2619563921         Manu_QLF          986

2. Retweets

In [54]:
all_users_df.sort_values('retweet_count', ascending=False).reset_index()[:100]

index              user_id        user_name follower_count  \
0    106793            138372303     FIFAWorldCup       11173420   
1   1055790             44409004          shakira       53305430   
2    296818             18091004           ESPNFC        6958316   
3    329417            330262748   FabrizioRomano       12796954   
4   1073074             44196397         elonmusk      122510051   
5    257908           1561123663       brfootball        5967973   
6    356251  1138458175663988738          PopBase         467303   
7     54114            119593082        FOXSoccer        1436531   
8    354605            627673190  ChampionsLeague       45779558   
9    488317              2557521             espn       43696497   
10   780824           4429003533         PopCrave        1202948   
11  1316018   805532293951606785   MattWallace888         786600   
12  1019814   973338313661861889  btschartsdailys         726344   
13   344665           4766665333      Culturecams          40529   
14  1710867             14123683      andy_murray        3563717   
15    34008           1580762162  chidajunsei3163         103445   
16   358568             26809005             goal        7149176   
17  1472143            266135518              KSI        8560920   
18   229445             69962890   2010MisterChip        3478499   
19   335777               890891   BleacherReport       13219827   
20   238527             27489253     AVFCOfficial        2206526   
21  1627288             57606154    OfficialBHAFC         712265   
22  1006820           2192335506     TeamCRonaldo         287111   
23  1643404             43152482    AlejandroSanz       19800789   
24   645890           3324689194   TheEuropeanLad         468941   
25   306635             89452962           ESPNUK         655573   
26    42996             39692424     AlertaNews24        1225927   
27   854770  1447233497731674117  AlbicelesteTalk         152174   
28  1729749  1121682547652513792  tanawatofficial         422348   
29   102523            385473825          Squawka        1219522   
30    84613            462312323  Football__Tweet         760537   
31   476556           2228960582  ScottLEOWarrior         446240   
32   141738            435225922       sportbible        1835349   
33  1559476            216299334      piersmorgan        8337074   
34    86532  1419145241018445825       _FaridKhan          38300   
35  1179701             26257166     SportsCenter       41400044   
36   277070            571964518    TrollFootball        2417482   
37   148439           2333988375      lnstantFoot        1501308   
38    65219   896550698543874049          UTDTrey         391678   
39   141168            265902729         BBCSport       10631968   
40   206942  1269288366954237952      mocnknights            679   
41  1662041             14130366     sundarpichai        5054964   
42  1671863  1049334054728949760   magicshopbooks          21555   
43   448835  1526619156669292544  Sports_Himanshu           2348   
44   481155             34613288          Arsenal       20756694   
45  1103847  1598975488978227202    FootballApeFC          19609   
46  1505537  1520612280643162112    DashDobrofsky         160402   
47  1387028           3437532637  myogiadityanath       22984910   
48   595911            427089628       lexfridman        2354327   
49   158412             14387275       paddypower         693633   
50  1413730            121002462      SC_Griffith           8258   
51   601606            533085085              M10       26308719   
52   173067  1272522535163355140        CFCDUBois         260580   
53  1004143            279581853     AchrafHakimi        1473974   
54   229197             44606764          OptaJoe        1334253   
55  1187416            352315445       minnielmin            457   
56  1801263            135421739        sachin_rt       38256550   
57  1335015  1045732637527011328       Aichalamet           12

3. Overall number of tweets posted

In [55]:
all_users_df.sort_values('tweet_count', ascending=False).reset_index()[:100]

index              user_id        user_name follower_count  \
0     16234  1489397690551386112     fifawc_ABEMA         177493   
1    648237           1183565334           lilo6y          53800   
2    649587           1375927867       98coolnola          37857   
3     14498             16946704       SBSSportau          55182   
4    307181  1258241001048739840     RishabhDeo11              1   
5    164996            349270821        Lared1061         150813   
6    439134  1563791149898338306          moanow1              4   
7    418795  1373313620327559172      EspnAllGame             54   
8    685694  1487425884378497027  Krishna78494082              9   
9    711195  1439834791168282629        sahu_cseb              8   
10   422778             72098300  sportsnewsreal2             33   
11   732678   839408617753309184     KhudeShubham              8   
12    60009           1132089344         Temba_qa           1607   
13   744399  1384200679145623553  IgTTNtla01b6jy1              0   
14   311660  1569187904546816001       FIFALIVE72             61   
15   423169             55162420  planetstreams99             26   
16   692608  1593671447972663296          KoyceWx              1   
17   422968             94947537    hdstreamslink             35   
18   307254  1535131918932783104  FootballLive710            143   
19   641902  1563791153404715008           wobcal              2   
20   217547   964541588746768384   Michel64819930             61   
21   307210  1554110239116443648       live_scor2             37   
22   751791  1583106271313424384   MDROKY35308626             19   
23   164154  1500803205189410818      SakidTahmid              1   
24   424051             32339605  dreamboxstreams             26   
25   585458  1594705396157251584  SabbirR15722035              2   
26   650900  1554353771769118721          zerdil6              3   
27    90595  1593295797734473732      Kxw06526285              0   
28   423458             62484823     a2zstreamstv             29   
29   252577            261792322     HancokUndead           1712   
30   171997  1191053322863730689       koora_lite           1121   
31   704170   888963967137271808  jagdish24067585              0   
32   171969   921651073336537088         tamil_is              9   
33     2213  1579998556026724352     IranZiba1401             31   
34   308097           3255473401            ABEMA        2102036   
35   994026  1295370538106277889  Rahulsh87550783              5   
36   606382  1592172490478223360       NrgSports4             44   
37    57375           4605538581        sir_Khaay           2684   
38   601905            189698435   watchnowfreetv            179   
39   175407           1345044672     momentof1994            602   
40   423698             24235729  planetstreamshq             33   
41   482610   839172484402458628  Liyakat96996251              1   
42   168894  1588941229609213952   Parvez39166996              0   
43   418785           3308820259     JanaProsanta            112   
44   418790  1268136636623147008   vivekrajgajula              2   
45   429734   834301604748746752  AJAYKUM02767038              1   
46   458455  1597928554997481472         Razib210              1   
47    51016  1535128154032943104         BoxingI6            353   
48   475025  1330510451860451329    HuzayfaFreund              2   
49    83280  1578055718904639488  RobertSitinjak5             12   
50   494640           2996611142      kailashp112              1   
51    32212           3402582670     stephenoflyf           4870   
52    50719   817355066629713920  StarMikazuki922           2967   
53   171849  1480254798469472257  IamAfzalSolangi             15   
54   719090  1602326148116930561         onionblu              0   
55   415030  1561387170908483587      rugbylivepc            118   
56    85576  1596352910866014208   BomBom65768245              0   
57  1420599  1594234416028680193         DemiriWz             

4. Follower count

In [56]:
all_users_df.sort_values('follower_count', ascending=False).reset_index()[:100]

index     user_id        user_name follower_count following_count  \
0   1073074    44196397         elonmusk      122510051             158   
1    325228    10228272          YouTube       77936208            1195   
2   1347020      428333           cnnbrk       64113175             122   
3    315959      759251              CNN       60721570            1092   
4    284253      807095          nytimes       54691282             863   
5   1055790    44409004          shakira       53305430             234   
6   1427961    23083404        KingJames       52558947             184   
7    498429     5402612      BBCBreaking       51504224               3   
8    355405    14872237       realmadrid       45853490              65   
9    354605   627673190  ChampionsLeague       45779558             628   
10   366558    96951800      FCBarcelona       45278284             111   
11   488317     2557521             espn       43696497             430   
12  1179701    26257166     SportsCenter       41400044             732   
13   680139      742143         BBCWorld       39162190              19   
14  1801263   135421739        sachin_rt       38256550              81   
15   384471   343627165    premierleague       37699137              82   
16   192620   558797310           ManUtd       34070923             174   
17   672001     5988062     TheEconomist       27135889             151   
18   601606   533085085              M10       26308719             172   
19    28496     1652541          Reuters       25645747            1236   
20  1528912    31927467          pitbull       24969824            2299   
21   421732    22910295        ChelseaFC       23397384             303   
22   313484     1367531          FoxNews       23297602             262   
23  1419836    92724677   virendersehwag       23152722             139   
24  1387028  3437532637  myogiadityanath       22984910              52   
25   176966    19583545              LFC       22907135          357383   
26     7749    33884545            CNNEE       22289897             416   
27  1033542    29479000     10Ronaldinho       21373341             144   
28   481155    34613288          Arsenal       20756694          145485   
29   157477   112764971   FCBarcelona_es       20548713             104   
30   293993     3108351              WSJ       20346107            1077   
31   166559     2467791   washingtonpost       19934909            1705   
32  1643404    43152482    AlejandroSanz       19800789            4433   
33    22724    69183155         detikcom       19467984              35   
34   228253    14293310             TIME       19397785             531   
35   171291    42606652           aajtak       19244016              35   
36   521680    91478624           Forbes       18569259            5058   
37  1771407    28785486              ABC       17763762             447   
38   151905    37034483             ndtv       17628355              15   
39  1078745    14213711    DaniloGentili       17253320             413   
40   505303   285332860      WayneRooney       17228039             546   
41    50867   148711787        AlArabiya       16216014              48   
42    84268    51241574               AP       16024656            6725   
43    79532   140070953          FIFAcom       15796302             268   
44   355955   452155423     realmadriden       15795081              63   
45  1189732   278235826         elissakh       15777591             401   
46   697250   300392950    TwitterSports       14836477            2141   
47   206108      612473          BBCNews       14815220              97   
48    34139   134758540     timesofindia       14632555             320   
49   503758    14573900          ManCity       14477129           49604   
50  1436095   295026890         SonuSood       14406079             171   
51   285402     2097571             cnni       14307716             368   
52   397016    60327493        

### ENTITIES

OK - now, let's have a look at the Twitter-supplied entities. Luckily, we've already recorded aggregate figures for this while initially collecting the data, so we needn't iterate over all tweets again. 

We will start by looking at a similar league-table approach, but then select some potentially more *interesting* / *unexpected* entitity names, with the goal of then producing subsamples of tweets which were classified by Twitter as being part of those entities. This may provide an interesting base for further analysis.  

In [61]:
entities_files = os.listdir(ENTITIES_PATH)
entities_files.sort()

Let's build a massive dict of entities with counts and timestamps... 

In [101]:
entities_by_time_df = pd.DataFrame()
entities_total = {}

In [102]:
for file in entities_files:
    with open(ENTITIES_PATH+file, 'r') as infile:
        timestamp = pd.to_datetime(file[8:27].replace('-', ' ').replace('_', '-'))
        tmp = json.load(infile)
        
        # let's start with overall
        for k,v in tmp.items():
            if k not in entities_total.keys():
                entities_total[k] = v
            else:
                entities_total[k] += v

        # now the df
        tmp_df = pd.DataFrame([tmp]).transpose().reset_index().rename(columns={'index' : 'entity', 0 : 'freq'})
        tmp_df['timestamp'] = timestamp
        entities_by_time_df = pd.concat([entities_by_time_df, tmp_df], axis=0)

In [103]:
entities_total

{'Unified Twitter Taxonomy': 2400448,
 'Business Taxonomy': 2392712,
 'Sports League': 2291172,
 'Events [Entity Service]': 2146606,
 'Sports Event': 2129114,
 'Sports Series': 2106010,
 'Sport': 1834803,
 'Sports Team': 1378975,
 'Person': 932491,
 'Soccer Match': 1451408,
 'Brand': 257439,
 'Athlete': 701230,
 'Entities [Entity Service]': 185872,
 'Musician': 107604,
 'Music Genre': 93105,
 'Interests and Hobbies Category': 74254,
 'Brand Vertical': 142029,
 'TV Shows': 81116,
 'Interests and Hobbies': 33057,
 'Fan Community': 10442,
 'Politician': 40555,
 'Interests and Hobbies Vertical': 53614,
 'Product': 66406,
 'Place': 11290,
 'Technology': 24655,
 'Cities': 7343,
 'Coach': 15872,
 'Sports Personality': 39064,
 'News Vertical': 13670,
 'Video Game': 45290,
 'Journalist': 48735,
 'Video Game Hardware': 6034,
 'Digital Assets & Crypto': 4501,
 'Basketball Game': 1363,
 'Entertainment Personality': 21010,
 'Political Body': 5483,
 'Product Taxonomy': 2635,
 'Global TV Show': 2586,

#### EXPORTS

In [ ]:
all_users_df
prolific_users_df